cpp/src/strings/convert/convert_urls.cu

```diff
@@ -222,30 +331,87 @@ std::unique_ptr<column> url_decode(
   size_type strings_count = strings.size();
   if (strings_count == 0) return make_empty_strings_column(stream, mr);
 
-  auto strings_column = column_device_view::create(strings.parent(), stream);
-  auto d_strings      = *strings_column;
+  auto offset_count = strings_count + 1;
+  auto d_offsets    = strings.offsets().data<size_type>() + strings.offset();
+  auto d_in_chars   = strings.chars().data<char>();
+  // determine index of first character in base column
+  size_type chars_start = (strings.offset() == 0) ? 0
+                                                  : cudf::detail::get_value<int32_t>(
+                                                      strings.offsets(), strings.offset(), stream);
+  size_type chars_end = (offset_count == strings.offsets().size())
+                          ? strings.chars_size()
+                          : cudf::detail::get_value<int32_t>(
+                              strings.offsets(), strings.offset() + strings_count, stream);
+  size_type chars_bytes = chars_end - chars_start;
+
+  url_decode_escape_detector esc_detector{device_span<char const>(d_in_chars, chars_end)};
+
+  // Count the number of URL escape sequences across all strings, ignoring string boundaries.
+  // This may count more sequences than actually are there since string boundaries are ignored.
+  size_type esc_count = thrust::count_if(rmm::exec_policy(stream),
+                                         thrust::make_counting_iterator<size_type>(chars_start),
+                                         thrust::make_counting_iterator<size_type>(chars_end),
+                                         esc_detector);
+
+  if (esc_count == 0) {
+    // nothing to replace, so just copy the input column
+    return std::make_unique<cudf::column>(strings.parent());
+  }
+
+  // create a vector of the potential escape sequence positions
+  rmm::device_uvector<size_type> esc_positions(esc_count, stream);
+  auto d_esc_positions = esc_positions.data();
+  thrust::copy_if(rmm::exec_policy(stream),
+                  thrust::make_counting_iterator<size_t>(chars_start),
+                  thrust::make_counting_iterator<size_t>(chars_end),
+                  d_esc_positions,
+                  esc_detector);
+
+  // In-place remove any escape positions that crossed string boundaries.
+  device_span<int32_t const> d_offsets_span(d_offsets, offset_count);
+  auto esc_pos_end = thrust::remove_if(rmm::exec_policy(stream),
+                                       d_esc_positions,
+                                       d_esc_positions + esc_count,
+                                       url_decode_esc_position_filter{d_offsets_span});
```

Comment:
If you want to avoid declaring a variable:
```suggestion
                                       url_decode_esc_position_filter{{d_offsets, offset_count}});
```

Reply:
`d_offsets_span` is being reused for its `.begin()` and `.end()` methods below this, so I'd prefer to keep the variable unless you would rather see `d_offsets` and `d_offsets + offset_count` used explicitly instead.